File: src/frontend/cte/lib/app/data/model/cte_aquaviario_balsa_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:cte/app/data/model/model_imports.dart';
import 'package:cte/app/data/domain/domain_imports.dart';

class CteAquaviarioBalsaModel {
	int? id;
	int? idCteAquaviario;
	String? idBalsa;
	int? numeroViagem;
	String? direcao;
	String? portoEmbarque;
	String? portoTransbordo;
	String? portoDestino;
	String? tipoNavegacao;
	String? irin;
	CteAquaviarioModel? cteAquaviarioModel;

	CteAquaviarioBalsaModel({
		this.id,
		this.idCteAquaviario,
		this.idBalsa,
		this.numeroViagem,
		this.direcao,
		this.portoEmbarque,
		this.portoTransbordo,
		this.portoDestino,
		this.tipoNavegacao,
		this.irin,
		this.cteAquaviarioModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'id_balsa',
		'numero_viagem',
		'direcao',
		'porto_embarque',
		'porto_transbordo',
		'porto_destino',
		'tipo_navegacao',
		'irin',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Id Balsa',
		'Numero Viagem',
		'Direcao',
		'Porto Embarque',
		'Porto Transbordo',
		'Porto Destino',
		'Tipo Navegacao',
		'Irin',
	];

	CteAquaviarioBalsaModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idCteAquaviario = jsonData['idCteAquaviario'];
		idBalsa = jsonData['idBalsa'];
		numeroViagem = jsonData['numeroViagem'];
		direcao = CteAquaviarioBalsaDomain.getDirecao(jsonData['direcao']);
		portoEmbarque = jsonData['portoEmbarque'];
		portoTransbordo = jsonData['portoTransbordo'];
		portoDestino = jsonData['portoDestino'];
		tipoNavegacao = CteAquaviarioBalsaDomain.getTipoNavegacao(jsonData['tipoNavegacao']);
		irin = jsonData['irin'];
		cteAquaviarioModel = jsonData['cteAquaviarioModel'] == null ? CteAquaviarioModel() : CteAquaviarioModel.fromJson(jsonData['cteAquaviarioModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idCteAquaviario'] = idCteAquaviario != 0 ? idCteAquaviario : null;
		jsonData['idBalsa'] = idBalsa;
		jsonData['numeroViagem'] = numeroViagem;
		jsonData['direcao'] = CteAquaviarioBalsaDomain.setDirecao(direcao);
		jsonData['portoEmbarque'] = portoEmbarque;
		jsonData['portoTransbordo'] = portoTransbordo;
		jsonData['portoDestino'] = portoDestino;
		jsonData['tipoNavegacao'] = CteAquaviarioBalsaDomain.setTipoNavegacao(tipoNavegacao);
		jsonData['irin'] = irin;
		jsonData['cteAquaviarioModel'] = cteAquaviarioModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idCteAquaviario = plutoRow.cells['idCteAquaviario']?.value;
		idBalsa = plutoRow.cells['idBalsa']?.value;
		numeroViagem = plutoRow.cells['numeroViagem']?.value;
		direcao = plutoRow.cells['direcao']?.value != '' ? plutoRow.cells['direcao']?.value : 'AAA';
		portoEmbarque = plutoRow.cells['portoEmbarque']?.value;
		portoTransbordo = plutoRow.cells['portoTransbordo']?.value;
		portoDestino = plutoRow.cells['portoDestino']?.value;
		tipoNavegacao = plutoRow.cells['tipoNavegacao']?.value != '' ? plutoRow.cells['tipoNavegacao']?.value : 'AAA';
		irin = plutoRow.cells['irin']?.value;
		cteAquaviarioModel = CteAquaviarioModel();
		cteAquaviarioModel?.numeroControle = plutoRow.cells['cteAquaviarioModel']?.value;
	}	

	CteAquaviarioBalsaModel clone() {
		return CteAquaviarioBalsaModel(
			id: id,
			idCteAquaviario: idCteAquaviario,
			idBalsa: idBalsa,
			numeroViagem: numeroViagem,
			direcao: direcao,
			portoEmbarque: portoEmbarque,
			portoTransbordo: portoTransbordo,
			portoDestino: portoDestino,
			tipoNavegacao: tipoNavegacao,
			irin: irin,
		);			
	}

	
}

File: backend/cte/src/model/cte_aquaviario_balsa_model.py
from src import db
from src.model.cte_aquaviario_model import CteAquaviarioModel


class CteAquaviarioBalsaModel(db.Model):
    __tablename__ = 'cte_aquaviario_balsa'

    id = db.Column(db.Integer, primary_key=True)
    id_balsa = db.Column(db.String(60))
    numero_viagem = db.Column(db.Integer)
    direcao = db.Column(db.String(1))
    porto_embarque = db.Column(db.String(60))
    porto_transbordo = db.Column(db.String(60))
    porto_destino = db.Column(db.String(60))
    tipo_navegacao = db.Column(db.String(1))
    irin = db.Column(db.String(10))
    id_cte_aquaviario = db.Column(db.Integer, db.ForeignKey('cte_aquaviario.id'))

    cte_aquaviario_model = db.relationship('CteAquaviarioModel', foreign_keys=[id_cte_aquaviario])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_cte_aquaviario = data.get('idCteAquaviario')
        self.id_balsa = data.get('idBalsa')
        self.numero_viagem = data.get('numeroViagem')
        self.direcao = data.get('direcao')
        self.porto_embarque = data.get('portoEmbarque')
        self.porto_transbordo = data.get('portoTransbordo')
        self.porto_destino = data.get('portoDestino')
        self.tipo_navegacao = data.get('tipoNavegacao')
        self.irin = data.get('irin')

    def serialize(self):
        return {
            'id': self.id,
            'idCteAquaviario': self.id_cte_aquaviario,
            'idBalsa': self.id_balsa,
            'numeroViagem': self.numero_viagem,
            'direcao': self.direcao,
            'portoEmbarque': self.porto_embarque,
            'portoTransbordo': self.porto_transbordo,
            'portoDestino': self.porto_destino,
            'tipoNavegacao': self.tipo_navegacao,
            'irin': self.irin,
            'cteAquaviarioModel': self.cte_aquaviario_model.serialize() if self.cte_aquaviario_model else None,
        }